src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/Dependencies/Assets/AssetsFileDependenciesSnapshot.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using NuGet.Common;
+using NuGet.ProjectModel;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Tree.Dependencies.Assets
+{
+    /// <summary>
+    /// Immutable snapshot of data captured from <c>project.assets.json</c>.
+    /// </summary>
+    internal sealed class AssetsFileDependenciesSnapshot
+    {
+        private static readonly LockFileFormat s_lockFileFormat = new LockFileFormat();
+
+        public static AssetsFileDependenciesSnapshot Empty { get; } = new AssetsFileDependenciesSnapshot(null);
+
+        /// <summary>
+        /// List of diagnostic messages included in the snapshot. May be empty.
+        /// </summary>
+        public ImmutableArray<AssetsFileLogMessage> Logs { get; }
+
+        /// <summary>
+        /// Produces an updated snapshot by reading the <c>project.assets.json</c> file at <paramref name="path"/>.
+        /// If the file could not be read, or no changes are detected, the current snapshot (this) is returned.
+        /// </summary>
+        public AssetsFileDependenciesSnapshot UpdateFromAssetsFile(string? path)
+        {
+            try
+            {
+                // Parse the file
+                using var fileStream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, 4096 * 10, FileOptions.SequentialScan);
```

Comment:
I think you want to combine FileShare.Delete here to avoid blocking folks from deleting the file while you are reading it.

Reply:
Good idea. I never knew this was a `[Flags]` enum!